File: campus-navigate-app/flutter_app/lib/domain/pdr/algorithms/pdr_pipeline.dart
import 'dart:math' as math;

import 'package:indoor_navigation/domain/common/angle_math.dart';
import 'package:indoor_navigation/domain/common/javascript_number.dart';
import 'package:indoor_navigation/domain/config/app_config.dart';
import 'package:indoor_navigation/domain/navigation/derived_navigation_estimate.dart';
import 'package:indoor_navigation/domain/pdr/algorithms/heading_candidates.dart';
import 'package:indoor_navigation/domain/pdr/algorithms/movement_gate.dart';
import 'package:indoor_navigation/domain/pdr/algorithms/step_detection.dart';
import 'package:indoor_navigation/domain/pdr/algorithms/transient_motion_batch.dart';
import 'package:indoor_navigation/domain/pdr/pdr_models.dart';

PdrPipelineResult runPdrPipeline({
  PdrPipelineConfig config = defaultPdrPipelineConfig,
  required double desiredHeadingDegrees,
  required int nowMs,
  double? pixelsPerMeter,
  required PdrPipelineState previousState,
  required List<MotionInputSample> samples,
}) {
  final batch = createTransientMotionBatch(
    config: config,
    nowMs: nowMs,
    samples: samples,
  );
  final observedHeadingDegrees = batch.acceptedSamples.isEmpty
      ? previousState.headingDegrees
      : circularMeanDegrees(
          batch.acceptedSamples.map((sample) => sample.headingDegrees),
        );
  final headingCandidates = rankHeadingCandidates(
    desiredHeadingDegrees: desiredHeadingDegrees,
    observedHeadingDegrees: observedHeadingDegrees,
    previousHeadingDegrees: previousState.headingDegrees,
  );
  final rotationWindowState = _createNextRotationWindowState(
    config: config,
    nowMs: nowMs,
    observedHeadingDegrees: observedHeadingDegrees,
    previousState: previousState,
  );
  final rawStepDiagnostic = analyzeStepDetection(
    config: config,
    previousState: previousState,
    samples: batch.acceptedSamples,
  );
  final stepDiagnostic = _applyPhoneRotationGate(
    config: config,
    rotationHeadingTravelDegrees:
        rotationWindowState.rotationHeadingTravelDegrees,
    stepDiagnostic: rawStepDiagnostic,
  );
  final stepCount = stepDiagnostic.stepCount;
  final movementDirection = getRouteMovementDirection(
    config: config,
    desiredHeadingDegrees: desiredHeadingDegrees,
    observedHeadingDegrees: observedHeadingDegrees,
  );
  final movementHeading = movementDirection == RouteMovementDirection.backward
      ? normalizeDegrees(desiredHeadingDegrees + 180)
      : headingCandidates.firstOrNull?.headingDegrees ?? observedHeadingDegrees;
  final effectivePixelsPerMeter = pixelsPerMeter != null && pixelsPerMeter > 0
      ? pixelsPerMeter
      : config.fallbackPixelsPerMeter;
  final movementGuardState = _createNextMovementGuardState(
    config: config,
    movementDirection: movementDirection,
    nowMs: nowMs,
    observedHeadingDegrees: observedHeadingDegrees,
    previousState: previousState,
    stepDiagnostic: stepDiagnostic,
  );
  final movementBlockedReason = _getMovementBlockedReason(
    config: config,
    guardState: movementGuardState,
    movementDirection: movementDirection,
    nowMs: nowMs,
    previousState: previousState,
    stepDiagnostic: stepDiagnostic,
  );
  final distancePixels = stepCount > 0 && movementBlockedReason == null
      ? stepCount * config.stepLengthMeters * effectivePixelsPerMeter
      : 0.0;
  final headingRadians = movementHeading * math.pi / 180;
  final latestStepTimestampMs = getLatestStepTimestamp(
    samples: batch.acceptedSamples,
    stepCount: stepCount,
  );
  final nextState = PdrPipelineState(
    backwardConfirmationTimestampMs:
        _shouldStoreBackwardConfirmation(
          config: config,
          movementDirection: movementDirection,
          stepDiagnostic: stepDiagnostic,
        )
        ? latestStepTimestampMs ?? previousState.backwardConfirmationTimestampMs
        : previousState.backwardConfirmationTimestampMs,
    headingDegrees: observedHeadingDegrees,
    lastStepTimestampMs:
        latestStepTimestampMs ?? previousState.lastStepTimestampMs,
    rotationHeadingSnapshots: rotationWindowState.rotationHeadingSnapshots,
    rotationHeadingTravelDegrees:
        rotationWindowState.rotationHeadingTravelDegrees,
    shakeCooldownUntilMs: movementGuardState.shakeCooldownUntilMs,
    shakeSpikeCount: movementGuardState.shakeSpikeCount,
    shakeWindowStartedAtMs: movementGuardState.shakeWindowStartedAtMs,
    // The TypeScript next-state literal intentionally does not carry the
    // one-shot `startedAtMs` marker forward, so the optional field is omitted.
    timestampMs: nowMs,
    turnInPlaceUntilMs: movementGuardState.turnInPlaceUntilMs,
    x: previousState.x + math.cos(headingRadians) * distancePixels,
    y: previousState.y + math.sin(headingRadians) * distancePixels,
  );
  final latencyMs = _calculateLatencyMs(batch.acceptedSamples, nowMs);

  return PdrPipelineResult(
    acceptedSampleCount: batch.acceptedSamples.length,
    diagnostics: PdrPipelineDiagnostics(
      batch: PdrBatchDiagnostic(
        acceptedSampleCount: batch.acceptedSamples.length,
        batchWindowMs: config.batchWindowMs,
        droppedSampleCount: batch.droppedSampleCount,
        maxBatchAgeMs: config.maxBatchAgeMs,
        maxSamplesPerBatch: config.maxSamplesPerBatch,
        rawSampleCount: samples.length,
        sampleEndTimestampMs: batch.acceptedSamples.lastOrNull?.timestampMs,
        sampleStartTimestampMs: batch.acceptedSamples.firstOrNull?.timestampMs,
      ),
      configSnapshot: config,
      heading: PdrHeadingDiagnostic(
        desiredHeadingDegrees: desiredHeadingDegrees,
        observedHeadingDegrees: observedHeadingDegrees,
        previousHeadingDegrees: previousState.headingDegrees,
        topCandidate: headingCandidates.firstOrNull,
      ),
      latencyMs: latencyMs.toDouble(),
      movement: PdrMovementDiagnostic(
        blockedReason: movementBlockedReason,
        direction: movementDirection,
        distancePixels: javascriptToFixedNumber(distancePixels, 3),
        headingDegrees: movementHeading,
        movedStepCount: distancePixels > 0 ? stepCount : 0,
        pixelsPerMeter: javascriptToFixedNumber(effectivePixelsPerMeter, 3),
        stepLengthMeters: config.stepLengthMeters,
      ),
      step: stepDiagnostic,
    ),
    droppedSampleCount: batch.droppedSampleCount,
    estimate: DerivedNavigationEstimate(
      confidence: _calculateConfidence(
        batch.acceptedSamples.length,
        headingCandidates.firstOrNull?.score ?? 0,
      ),
      headingDegrees: observedHeadingDegrees,
      source: DerivedNavigationEstimateSource.pdrSummary,
      timestampMs: nowMs,
      x: nextState.x,
      y: nextState.y,
    ),
    headingCandidates: headingCandidates,
    latencyMs: latencyMs.toDouble(),
    nextState: nextState,
  );
}

double _calculateConfidence(int sampleCount, double headingScore) {
  final sampleConfidence = math.min(1, sampleCount / 6);
  return javascriptToFixedNumber(
    sampleConfidence * 0.35 + headingScore * 0.65,
    3,
  );
}

_RotationWindowState _createNextRotationWindowState({
  required PdrPipelineConfig config,
  required int nowMs,
  required double observedHeadingDegrees,
  required PdrPipelineState previousState,
}) {
  final snapshots =
      <PdrHeadingSnapshot>[
            ...?previousState.rotationHeadingSnapshots,
            PdrHeadingSnapshot(
              headingDegrees: observedHeadingDegrees,
              timestampMs: nowMs,
            ),
          ]
          .where(
            (snapshot) =>
                nowMs - snapshot.timestampMs <= config.rotationOnlyWindowMs,
          )
          .toList(growable: false);

  return _RotationWindowState(
    rotationHeadingSnapshots: snapshots,
    rotationHeadingTravelDegrees: _calculateHeadingTravelDegrees(snapshots),
  );
}

double _calculateHeadingTravelDegrees(List<PdrHeadingSnapshot> snapshots) {
  var travelDegrees = 0.0;
  for (var index = 1; index < snapshots.length; index += 1) {
    travelDegrees += shortestAngleDistanceDegrees(
      snapshots[index].headingDegrees,
      snapshots[index - 1].headingDegrees,
    );
  }
  return javascriptToFixedNumber(travelDegrees, 3);
}

StepDetectionDiagnostic _applyPhoneRotationGate({
  required PdrPipelineConfig config,
  required double rotationHeadingTravelDegrees,
  required StepDetectionDiagnostic stepDiagnostic,
}) {
  final isPhoneRotation =
      stepDiagnostic.rejectReason == StepRejectReason.accepted &&
      rotationHeadingTravelDegrees >= config.rotationOnlyHeadingTravelDegrees &&
      stepDiagnostic.averageAcceleration <=
          config.rotationOnlyMaxAverageAcceleration;

  return StepDetectionDiagnostic(
    averageAcceleration: stepDiagnostic.averageAcceleration,
    minAcceleration: stepDiagnostic.minAcceleration,
    peakAcceleration: stepDiagnostic.peakAcceleration,
    peakTimestampMs: stepDiagnostic.peakTimestampMs,
    rejectReason: isPhoneRotation
        ? StepRejectReason.phoneRotation
        : stepDiagnostic.rejectReason,
    rotationHeadingTravelDegrees: rotationHeadingTravelDegrees,
    stepCount: isPhoneRotation ? 0 : stepDiagnostic.stepCount,
    timeSinceLastStepMs: stepDiagnostic.timeSinceLastStepMs,
  );
}

MovementBlockedReason? _getMovementBlockedReason({
  required PdrPipelineConfig config,
  required _MovementGuardState guardState,
  required RouteMovementDirection movementDirection,
  required int nowMs,
  required PdrPipelineState previousState,
  required StepDetectionDiagnostic stepDiagnostic,
}) {
  if (stepDiagnostic.stepCount == 0) {
    return null;
  }
  final startedAtMs = previousState.startedAtMs;
  if (startedAtMs != null &&
      nowMs - startedAtMs < config.startupMovementLockMs) {
    return MovementBlockedReason.startupLock;
  }
  final shakeCooldownUntilMs = guardState.shakeCooldownUntilMs;
  if (shakeCooldownUntilMs != null && nowMs < shakeCooldownUntilMs) {
    return MovementBlockedReason.shakeCooldown;
  }
  final turnInPlaceUntilMs = guardState.turnInPlaceUntilMs;
  if (turnInPlaceUntilMs != null && nowMs < turnInPlaceUntilMs) {
    return MovementBlockedReason.turningInPlace;
  }
  if (movementDirection == RouteMovementDirection.blocked) {
    return MovementBlockedReason.heading;
  }
  if (movementDirection == RouteMovementDirection.backward &&
      stepDiagnostic.peakAcceleration < config.backwardMovementPeakThreshold) {
    return MovementBlockedReason.backwardWeakStep;
  }
  if (movementDirection == RouteMovementDirection.backward &&
      !_hasRecentBackwardConfirmation(
        config: config,
        peakTimestampMs: stepDiagnostic.peakTimestampMs,
        previousBackwardConfirmationTimestampMs:
            previousState.backwardConfirmationTimestampMs,
      )) {
    return MovementBlockedReason.backwardConfirming;
  }
  return null;
}

_MovementGuardState _createNextMovementGuardState({
  required PdrPipelineConfig config,
  required RouteMovementDirection movementDirection,
  required int nowMs,
  required double observedHeadingDegrees,
  required PdrPipelineState previousState,
  required StepDetectionDiagnostic stepDiagnostic,
}) {
  final previousCooldownUntilMs = previousState.shakeCooldownUntilMs;
  var shakeCooldownUntilMs =
      previousCooldownUntilMs != null && nowMs < previousCooldownUntilMs
      ? previousCooldownUntilMs
      : null;
  var shakeSpikeCount = previousState.shakeSpikeCount ?? 0;
  var shakeWindowStartedAtMs = previousState.shakeWindowStartedAtMs;
  final previousTurnInPlaceUntilMs = previousState.turnInPlaceUntilMs;
  var turnInPlaceUntilMs =
      previousTurnInPlaceUntilMs != null && nowMs < previousTurnInPlaceUntilMs
      ? previousTurnInPlaceUntilMs
      : null;

  if (shakeWindowStartedAtMs != null &&
      nowMs - shakeWindowStartedAtMs > config.shakeCooldownWindowMs) {
    shakeSpikeCount = 0;
    shakeWindowStartedAtMs = null;
  }

  if (stepDiagnostic.rejectReason == StepRejectReason.shakeTooHigh &&
      stepDiagnostic.peakTimestampMs != null) {
    final peakTimestampMs = stepDiagnostic.peakTimestampMs!;
    if (shakeWindowStartedAtMs == null ||
        peakTimestampMs - shakeWindowStartedAtMs >
            config.shakeCooldownWindowMs) {
      shakeWindowStartedAtMs = peakTimestampMs;
      shakeSpikeCount = 1;
    } else {
      shakeSpikeCount += 1;
    }

    if (shakeSpikeCount >= config.shakeCooldownTriggerCount) {
      shakeCooldownUntilMs = _maxInt(
        shakeCooldownUntilMs ?? 0,
        peakTimestampMs + config.shakeCooldownMs,
      );
      shakeWindowStartedAtMs = peakTimestampMs;
      shakeSpikeCount = 0;
    }
  }

  if (movementDirection != RouteMovementDirection.backward &&
      stepDiagnostic.stepCount > 0 &&
      shortestAngleDistanceDegrees(
            observedHeadingDegrees,
            previousState.headingDegrees,
          ) >=
          config.turnInPlaceHeadingDeltaDegrees) {
    turnInPlaceUntilMs = _maxInt(
      turnInPlaceUntilMs ?? 0,
      nowMs + config.turnInPlaceCooldownMs,
    );
  }

  return _MovementGuardState(
    shakeCooldownUntilMs: shakeCooldownUntilMs,
    shakeSpikeCount: shakeSpikeCount,
    shakeWindowStartedAtMs: shakeWindowStartedAtMs,
    turnInPlaceUntilMs: turnInPlaceUntilMs,
  );
}

bool _shouldStoreBackwardConfirmation({
  required PdrPipelineConfig config,
  required RouteMovementDirection movementDirection,
  required StepDetectionDiagnostic stepDiagnostic,
}) {
  return stepDiagnostic.stepCount > 0 &&
      movementDirection == RouteMovementDirection.backward &&
      stepDiagnostic.peakAcceleration >= config.backwardMovementPeakThreshold;
}

bool _hasRecentBackwardConfirmation({
  required PdrPipelineConfig config,
  required int? peakTimestampMs,
  required int? previousBackwardConfirmationTimestampMs,
}) {
  if (peakTimestampMs == null ||
      previousBackwardConfirmationTimestampMs == null) {
    return false;
  }
  return peakTimestampMs - previousBackwardConfirmationTimestampMs <=
      config.backwardConfirmationWindowMs;
}

int _calculateLatencyMs(List<MotionInputSample> samples, int nowMs) {
  final newestSample = samples.lastOrNull;
  return newestSample == null ? 0 : nowMs - newestSample.timestampMs;
}

int _maxInt(int left, int right) => left > right ? left : right;

final class _RotationWindowState {
  _RotationWindowState({
    required List<PdrHeadingSnapshot> rotationHeadingSnapshots,
    required this.rotationHeadingTravelDegrees,
  }) : rotationHeadingSnapshots = List.unmodifiable(rotationHeadingSnapshots);

  final List<PdrHeadingSnapshot> rotationHeadingSnapshots;
  final double rotationHeadingTravelDegrees;
}

final class _MovementGuardState {
  const _MovementGuardState({
    required this.shakeCooldownUntilMs,
    required this.shakeSpikeCount,
    required this.shakeWindowStartedAtMs,
    required this.turnInPlaceUntilMs,
  });

  final int? shakeCooldownUntilMs;
  final int shakeSpikeCount;
  final int? shakeWindowStartedAtMs;
  final int? turnInPlaceUntilMs;
}

File: tarumt-nav-app/expo-app/src/pdr/model/pdrPipeline.ts
import type {
  MotionInputSample,
  PdrPipelineConfig,
  PdrPipelineResult,
  PdrPipelineState,
  StepDetectionDiagnostic,
} from '../type';
import { circularMeanDegrees, shortestAngleDistanceDegrees } from './angleModel';
import { createTransientMotionBatch } from './batchModel';
import { DEFAULT_PDR_PIPELINE_CONFIG } from './pdrConfig';
import { rankHeadingCandidates } from './headingRankModel';
import { getRouteMovementDirection } from './movementGateModel';
import { analyzeStepDetection, getLatestStepTimestamp } from './stepDetectionModel';
import { normalizeDegrees } from './angleModel';

export function runPdrPipeline(input: {
  config?: Partial<PdrPipelineConfig>;
  desiredHeadingDegrees: number;
  nowMs: number;
  pixelsPerMeter?: number;
  previousState: PdrPipelineState;
  samples: readonly MotionInputSample[];
}): PdrPipelineResult {
  const config = {
    ...DEFAULT_PDR_PIPELINE_CONFIG,
    ...input.config,
  };
  const batch = createTransientMotionBatch({
    config,
    nowMs: input.nowMs,
    samples: input.samples,
  });
  const observedHeadingDegrees =
    batch.acceptedSamples.length === 0
      ? input.previousState.headingDegrees
      : circularMeanDegrees(
          batch.acceptedSamples.map((sample) => sample.headingDegrees),
        );
  const headingCandidates = rankHeadingCandidates({
    desiredHeadingDegrees: input.desiredHeadingDegrees,
    observedHeadingDegrees,
    previousHeadingDegrees: input.previousState.headingDegrees,
  });
  const rotationWindowState = createNextRotationWindowState({
    config,
    nowMs: input.nowMs,
    observedHeadingDegrees,
    previousState: input.previousState,
  });
  const rawStepDiagnostic = analyzeStepDetection({
    config,
    previousState: input.previousState,
    samples: batch.acceptedSamples,
  });
  const stepDiagnostic = applyPhoneRotationGate({
    config,
    rotationHeadingTravelDegrees:
      rotationWindowState.rotationHeadingTravelDegrees ?? 0,
    stepDiagnostic: rawStepDiagnostic,
  });
  const stepCount = stepDiagnostic.stepCount;
  const movementDirection = getRouteMovementDirection({
    config,
    desiredHeadingDegrees: input.desiredHeadingDegrees,
    observedHeadingDegrees,
  });
  const movementHeading =
    movementDirection === 'backward'
      ? normalizeDegrees(input.desiredHeadingDegrees + 180)
      : headingCandidates[0]?.headingDegrees ?? observedHeadingDegrees;
  const pixelsPerMeter =
    input.pixelsPerMeter && input.pixelsPerMeter > 0
      ? input.pixelsPerMeter
      : config.fallbackPixelsPerMeter;
  const movementGuardState = createNextMovementGuardState({
    config,
    movementDirection,
    nowMs: input.nowMs,
    observedHeadingDegrees,
    previousState: input.previousState,
    stepDiagnostic,
  });
  const movementBlockedReason = getMovementBlockedReason({
    config,
    guardState: movementGuardState,
    movementDirection,
    nowMs: input.nowMs,
    previousState: input.previousState,
    stepDiagnostic,
  });
  const distancePixels =
    stepCount > 0 && movementBlockedReason === null
      ? stepCount * config.stepLengthMeters * pixelsPerMeter
      : 0;
  const headingRadians = (movementHeading * Math.PI) / 180;
  const latestStepTimestampMs = getLatestStepTimestamp({
    samples: batch.acceptedSamples,
    stepCount,
  });
  const nextState = {
    headingDegrees: observedHeadingDegrees,
    lastStepTimestampMs:
      latestStepTimestampMs ?? input.previousState.lastStepTimestampMs,
    backwardConfirmationTimestampMs:
      shouldStoreBackwardConfirmation({
        config,
        movementDirection,
        stepDiagnostic,
      })
        ? latestStepTimestampMs ?? input.previousState.backwardConfirmationTimestampMs
        : input.previousState.backwardConfirmationTimestampMs,
    ...rotationWindowState,
    ...movementGuardState,
    timestampMs: input.nowMs,
    x: input.previousState.x + Math.cos(headingRadians) * distancePixels,
    y: input.previousState.y + Math.sin(headingRadians) * distancePixels,
  };

  const latencyMs = calculateLatencyMs(batch.acceptedSamples, input.nowMs);

  return {
    acceptedSampleCount: batch.acceptedSamples.length,
    diagnostics: {
      batch: {
        acceptedSampleCount: batch.acceptedSamples.length,
        batchWindowMs: config.batchWindowMs,
        droppedSampleCount: batch.droppedSampleCount,
        maxBatchAgeMs: config.maxBatchAgeMs,
        maxSamplesPerBatch: config.maxSamplesPerBatch,
        rawSampleCount: input.samples.length,
        sampleEndTimestampMs:
          batch.acceptedSamples[batch.acceptedSamples.length - 1]?.timestampMs ??
          null,
        sampleStartTimestampMs: batch.acceptedSamples[0]?.timestampMs ?? null,
      },
      configSnapshot: config,
      heading: {
        desiredHeadingDegrees: input.desiredHeadingDegrees,
        observedHeadingDegrees,
        previousHeadingDegrees: input.previousState.headingDegrees,
        topCandidate: headingCandidates[0] ?? null,
      },
      latencyMs,
      movement: {
        blockedReason: movementBlockedReason,
        direction: movementDirection,
        distancePixels: Number(distancePixels.toFixed(3)),
        headingDegrees: movementHeading,
        movedStepCount: distancePixels > 0 ? stepCount : 0,
        pixelsPerMeter: Number(pixelsPerMeter.toFixed(3)),
        stepLengthMeters: config.stepLengthMeters,
      },
      step: stepDiagnostic,
    },
    droppedSampleCount: batch.droppedSampleCount,
    estimate: {
      confidence: calculateConfidence(batch.acceptedSamples.length, headingCandidates[0]?.score ?? 0),
      headingDegrees: observedHeadingDegrees,
      source: 'pdr-summary',
      timestampMs: input.nowMs,
      x: nextState.x,
      y: nextState.y,
    },
    headingCandidates,
    latencyMs,
    nextState,
  };
}

function calculateConfidence(sampleCount: number, headingScore: number) {
  const sampleConfidence = Math.min(1, sampleCount / 6);
  return Number((sampleConfidence * 0.35 + headingScore * 0.65).toFixed(3));
}

function createNextRotationWindowState(input: {
  config: PdrPipelineConfig;
  nowMs: number;
  observedHeadingDegrees: number;
  previousState: PdrPipelineState;
}) {
  const previousSnapshots = input.previousState.rotationHeadingSnapshots ?? [];
  const snapshots = [
    ...previousSnapshots,
    {
      headingDegrees: input.observedHeadingDegrees,
      timestampMs: input.nowMs,
    },
  ].filter(
    (snapshot) =>
      input.nowMs - snapshot.timestampMs <= input.config.rotationOnlyWindowMs,
  );

  return {
    rotationHeadingSnapshots: snapshots,
    rotationHeadingTravelDegrees: calculateHeadingTravelDegrees(snapshots),
  };
}

function calculateHeadingTravelDegrees(
  snapshots: NonNullable<PdrPipelineState['rotationHeadingSnapshots']>,
) {
  let travelDegrees = 0;

  for (let index = 1; index < snapshots.length; index += 1) {
    const previousSnapshot = snapshots[index - 1];
    const snapshot = snapshots[index];
    if (!previousSnapshot || !snapshot) {
      continue;
    }

    travelDegrees += shortestAngleDistanceDegrees(
      snapshot.headingDegrees,
      previousSnapshot.headingDegrees,
    );
  }

  return Number(travelDegrees.toFixed(3));
}

function applyPhoneRotationGate(input: {
  config: PdrPipelineConfig;
  rotationHeadingTravelDegrees: number;
  stepDiagnostic: StepDetectionDiagnostic;
}): StepDetectionDiagnostic {
  if (
    input.stepDiagnostic.rejectReason !== 'ACCEPTED' ||
    input.rotationHeadingTravelDegrees <
      input.config.rotationOnlyHeadingTravelDegrees ||
    input.stepDiagnostic.averageAcceleration >
      input.config.rotationOnlyMaxAverageAcceleration
  ) {
    return {
      ...input.stepDiagnostic,
      rotationHeadingTravelDegrees: input.rotationHeadingTravelDegrees,
    };
  }

  return {
    ...input.stepDiagnostic,
    rejectReason: 'PHONE_ROTATION',
    rotationHeadingTravelDegrees: input.rotationHeadingTravelDegrees,
    stepCount: 0,
  };
}

function getMovementBlockedReason(input: {
  config: PdrPipelineConfig;
  guardState: {
    shakeCooldownUntilMs?: number;
    turnInPlaceUntilMs?: number;
  };
  movementDirection: ReturnType<typeof getRouteMovementDirection>;
  nowMs: number;
  previousState: PdrPipelineState;
  stepDiagnostic: ReturnType<typeof analyzeStepDetection>;
}):
  | 'backward-confirming'
  | 'backward-weak-step'
  | 'heading'
  | 'shake-cooldown'
  | 'startup-lock'
  | 'turning-in-place'
  | null {
  if (input.stepDiagnostic.stepCount === 0) {
    return null;
  }

  if (
    input.previousState.startedAtMs !== undefined &&
    input.nowMs - input.previousState.startedAtMs <
      input.config.startupMovementLockMs
  ) {
    return 'startup-lock';
  }

  if (
    input.guardState.shakeCooldownUntilMs !== undefined &&
    input.nowMs < input.guardState.shakeCooldownUntilMs
  ) {
    return 'shake-cooldown';
  }

  if (
    input.guardState.turnInPlaceUntilMs !== undefined &&
    input.nowMs < input.guardState.turnInPlaceUntilMs
  ) {
    return 'turning-in-place';
  }

  if (input.movementDirection === 'blocked') {
    return 'heading';
  }

  if (
    input.movementDirection === 'backward' &&
    input.stepDiagnostic.peakAcceleration <
      input.config.backwardMovementPeakThreshold
  ) {
    return 'backward-weak-step';
  }

  if (
    input.movementDirection === 'backward' &&
    !hasRecentBackwardConfirmation({
      config: input.config,
      peakTimestampMs: input.stepDiagnostic.peakTimestampMs,
      previousBackwardConfirmationTimestampMs:
        input.previousState.backwardConfirmationTimestampMs,
    })
  ) {
    return 'backward-confirming';
  }

  return null;
}

function createNextMovementGuardState(input: {
  config: PdrPipelineConfig;
  movementDirection: ReturnType<typeof getRouteMovementDirection>;
  nowMs: number;
  observedHeadingDegrees: number;
  previousState: PdrPipelineState;
  stepDiagnostic: ReturnType<typeof analyzeStepDetection>;
}) {
  const activeCooldownUntilMs =
    input.previousState.shakeCooldownUntilMs !== undefined &&
    input.nowMs < input.previousState.shakeCooldownUntilMs
      ? input.previousState.shakeCooldownUntilMs
      : undefined;
  let shakeCooldownUntilMs = activeCooldownUntilMs;
  let shakeSpikeCount = input.previousState.shakeSpikeCount ?? 0;
  let shakeWindowStartedAtMs = input.previousState.shakeWindowStartedAtMs;
  let turnInPlaceUntilMs =
    input.previousState.turnInPlaceUntilMs !== undefined &&
    input.nowMs < input.previousState.turnInPlaceUntilMs
      ? input.previousState.turnInPlaceUntilMs
      : undefined;

  if (
    shakeWindowStartedAtMs !== undefined &&
    input.nowMs - shakeWindowStartedAtMs > input.config.shakeCooldownWindowMs
  ) {
    shakeSpikeCount = 0;
    shakeWindowStartedAtMs = undefined;
  }

  if (
    input.stepDiagnostic.rejectReason === 'SHAKE_TOO_HIGH' &&
    input.stepDiagnostic.peakTimestampMs !== null
  ) {
    if (
      shakeWindowStartedAtMs === undefined ||
      input.stepDiagnostic.peakTimestampMs - shakeWindowStartedAtMs >
        input.config.shakeCooldownWindowMs
    ) {
      shakeWindowStartedAtMs = input.stepDiagnostic.peakTimestampMs;
      shakeSpikeCount = 1;
    } else {
      shakeSpikeCount += 1;
    }

    if (shakeSpikeCount >= input.config.shakeCooldownTriggerCount) {
      shakeCooldownUntilMs = Math.max(
        shakeCooldownUntilMs ?? 0,
        input.stepDiagnostic.peakTimestampMs + input.config.shakeCooldownMs,
      );
      shakeWindowStartedAtMs = input.stepDiagnostic.peakTimestampMs;
      shakeSpikeCount = 0;
    }
  }

  if (
    input.movementDirection !== 'backward' &&
    input.stepDiagnostic.stepCount > 0 &&
    shortestAngleDistanceDegrees(
      input.observedHeadingDegrees,
      input.previousState.headingDegrees,
    ) >= input.config.turnInPlaceHeadingDeltaDegrees
  ) {
    turnInPlaceUntilMs = Math.max(
      turnInPlaceUntilMs ?? 0,
      input.nowMs + input.config.turnInPlaceCooldownMs,
    );
  }

  return {
    shakeCooldownUntilMs,
    shakeSpikeCount,
    shakeWindowStartedAtMs,
    turnInPlaceUntilMs,
  };
}

function shouldStoreBackwardConfirmation(input: {
  config: PdrPipelineConfig;
  movementDirection: ReturnType<typeof getRouteMovementDirection>;
  stepDiagnostic: ReturnType<typeof analyzeStepDetection>;
}) {
  return (
    input.stepDiagnostic.stepCount > 0 &&
    input.movementDirection === 'backward' &&
    input.stepDiagnostic.peakAcceleration >=
      input.config.backwardMovementPeakThreshold
  );
}

function hasRecentBackwardConfirmation(input: {
  config: PdrPipelineConfig;
  peakTimestampMs: number | null;
  previousBackwardConfirmationTimestampMs?: number;
}) {
  if (
    input.peakTimestampMs === null ||
    input.previousBackwardConfirmationTimestampMs === undefined
  ) {
    return false;
  }

  return (
    input.peakTimestampMs - input.previousBackwardConfirmationTimestampMs <=
    input.config.backwardConfirmationWindowMs
  );
}

function calculateLatencyMs(
  samples: readonly MotionInputSample[],
  nowMs: number,
) {
  const newestSample = samples[samples.length - 1];
  return newestSample ? nowMs - newestSample.timestampMs : 0;
}
